### DISCUSSION TITLE
Postgres Json support

### DISCUSSION THREAD

**Discussion by alberk8** [linq2db/linq2db#4762] [Q&A]
I would like to know if postgres Jsonb column  is supported via Fluent Mapping?   I tried but keep getting casting error string to class.

```
//Table
public class Person 
{
   public string Name {get;set;}
   //Json column
   public Address Address {get;set;}
}


public class Address 
{
  public string Street {get;set;}
  public string PostCode {get;set}
}

```

**Comment by MaceWindu** [comment]
You need to assign `DataType=DataType.BinaryJson` to column

**Reply by alberk8** [reply]
does that mean i need to apply attribute to the POCO class?

**Reply by MaceWindu** [reply]
It could be done using fluent mapping

```cs
.Property(e => e.Address).HasDataType(DataType.BinaryJson)
```

Just noticed that you want to use POCO for json property. It require additional configuration right now. See this discussion how to setup npgsql and linq2db for it https://github.com/linq2db/linq2db/discussions/4738

**Reply by MaceWindu** [reply]
This example specifically https://github.com/linq2db/linq2db/discussions/4738#discussioncomment-11565738

**Reply by alberk8** [reply]
Thanks for the addition information.  I still does not follow in the link what needs to be done for read and write jsonb column.  Btw, is there an ongoing effort to get that fixed?

**Reply by MaceWindu** [reply]
1. you need to enable JSON support on npgsql side using `NpgsqlDataSourceBuilder` and pass this data source to linq2db:
```cs
var builder = new NpgsqlDataSourceBuilder(connectionString);
builder.EnableDynamicJson();
var ds = builder.Build();

using var db = new DataConnection(new DataOptions().UseConnectionFactory(CustomProvider.Instance, o => ds.CreateConnection()));
```

2. subclass database provider to implement POCO read (this is something we need to implement at some point on our side):
```cs
public override Expression GetReaderExpression(DbDataReader reader, int idx, Expression readerExpression, Type? toType)
	{
		var typeName = reader.GetDataTypeName(idx);

		if (typeName is "json" or "jsonb")
		{
			// return rd.GetFieldValue<ToType>(ordinal)
			return Expression.Lambda(
				Expression.Call(_rd, _getFieldValue.MakeGenericMethod(toType!), _ordinal),
				_rd, _ordinal);
		}
```
